src/wallet/multi_tracker.rs

```diff
@@ -0,0 +1,405 @@
+
+use bitcoin::{Network, OutPoint, Script, Transaction, TxOut, Txid};
+use miniscript::DescriptorTrait;
+
+#[allow(deprecated)]
+use crate::{
+    address_validator::AddressValidator,
+    database::{BatchDatabase, DatabaseUtils},
+    descriptor::{AsDerived, ExtendedDescriptor},
+    signer::{SignersContainer, TransactionSigner},
+    BlockTime, Error, KeychainKind, LocalUtxo, Wallet,
+};
+
+use super::{utils::SecpCtx, AddressInfo};
+
+/// Contains a descriptor and associated data
+#[derive(Debug, Clone)]
+pub struct DescriptorItem {
+    /// Wallet descriptor
+    pub descriptor: ExtendedDescriptor,
+    /// Keychain kind of descriptor - external/internal
+    pub keychain: KeychainKind,
+    /// Signers of descriptor
+    pub signers: Arc<SignersContainer>,
+}
+
+/// Contains a transaction and it's associated data
+#[derive(Debug, Clone)]
+pub struct TransactionItem {
+    /// Raw transaction
+    pub raw: Transaction,
+    /// Confirmed time (if any) of the transaction
+    pub confirmed: Option<BlockTime>,
+    /// Fees of the transaction (if avaliable)
+    pub fees: Option<u64>,
+}
+
+/// Contains a local tx output and it's associated data
+/// TODO @evanlinjin: Figure out if this is useful
+#[derive(Debug, Clone)]
+pub struct LocalOutputItem {
+    /// References this output
+    pub outpoint: OutPoint,
+    /// The actual tx output
+    pub txout: TxOut,
+    /// Confirmation time (if any) of the tx that contains this output
+    pub confirmed: Option<BlockTime>,
+    /// Path for `scriptPubKey` creation (descriptor_item, child_index)
+    pub script_path: (DescriptorItem, u32),
+    /// Satisfaction weight for the `scriptPubKey` (required weight of corresponding
+    /// `scriptSig`/`witness` data)
+    pub satisfaction_weight: u32,
+}
+
+/// Contains the "required" methods of [`MultiTracker`], where all other methods could be
+/// deried from (albiet probably not in an optimised manner).
+///
+/// TODO @evanlinjin: Should we have an `iter_txout` method as defined in `bdk_core`?
+/// <https://github.com/LLFourn/bdk_core_staging/blob/67d1e5499b0eb1de59f652c433c88942646593b3/bdk_core/src/descriptor_tracker.rs#L873>
+pub trait MultiTrackerInner {
+    /// Returns an iterator that ranges through all owned descriptors.
+    fn iter_descriptors(&self) -> Result<Box<dyn Iterator<Item = DescriptorItem> + '_>, Error>;
+
+    /// Returns an iterator that ranges through all owned signers.
+    fn iter_signers(
+        &self,
+    ) -> Result<Box<dyn Iterator<Item = Arc<dyn TransactionSigner>> + '_>, Error>;
+
+    /// Obtains owned UTXOs alongside their satisfaction weights.
+    ///
+    /// TODO @evanlinjin: We should have our own structure for Unspents.
+    /// Fields that make sense =>
+    /// * Confirmed height.
+    /// * Descriptor + Child index.
+    /// * Satisfaction weight of scriptPubKey (weight of corresponding scriptSig/witness data).
+    fn iter_unspent(&self) -> Result<Box<dyn Iterator<Item = (LocalUtxo, usize)> + '_>, Error>;
```

Comment:
However, `WeightedUtxo` can be of a foreign `Utxo` so it is not as "direct" as using an union here (since this method is about iterating unspent, and owned UTXOs). Should I keep it as is?